File: backend_samples/library/python/AzureAIVisionFaceDeviceAttestation/azure_ai_vision_face_deviceattestation/android/play_integrity_api.py
"""
Call Google's Play Integrity API to decode a client-supplied integrity token.

Reads service-account JSON from `GOOGLE_SERVICE_ACCOUNT_JSON` and the bundle id
from `ANDROID_PACKAGE_NAME`. Returns an :class:`IntegrityVerdictResult`.
"""

from __future__ import annotations

import json
import logging
import time
from dataclasses import dataclass
from typing import Optional

from ..config import get_attestation_config
from ..logging import track_dependency, track_event, track_exception

log = logging.getLogger("android.play_integrity_api")


@dataclass(frozen=True)
class IntegrityVerdictResult:
    """Outcome of a Play Integrity decode.

    ``tolerable`` is True only for the Google-unavailable (network / HTTP 5xx) or
    quota-exceeded (HTTP 429) conditions the
    ``allow_android_attestation_when_google_unavailable`` policy may accept; every
    other failure fails closed.
    """

    ok: bool
    verdict: Optional[dict] = None
    tolerable: bool = False
    reason: str = ""

    @staticmethod
    def success(verdict: dict) -> "IntegrityVerdictResult":
        return IntegrityVerdictResult(ok=True, verdict=verdict, reason="ok")

    @staticmethod
    def fail_tolerable(reason: str) -> "IntegrityVerdictResult":
        return IntegrityVerdictResult(ok=False, tolerable=True, reason=reason)

    @staticmethod
    def fail_hard(reason: str) -> "IntegrityVerdictResult":
        return IntegrityVerdictResult(ok=False, tolerable=False, reason=reason)


def _http_status(err: Exception) -> Optional[int]:
    """HTTP status code from a googleapiclient HttpError, else None."""
    try:
        from googleapiclient.errors import HttpError

        if isinstance(err, HttpError):
            status = getattr(getattr(err, "resp", None), "status", None)
            if status is not None:
                return int(status)
            status_code = getattr(err, "status_code", None)
            return int(status_code) if status_code is not None else None
    except Exception:
        pass
    return None


def _is_network_error(err: Exception) -> bool:
    """True for connectivity / timeout failures reaching Google."""
    # ConnectionError, TimeoutError, socket.timeout, socket.gaierror are OSError subclasses.
    if isinstance(err, OSError):
        return True
    for mod_name, cls_name in (("google.auth.exceptions", "TransportError"), ("httplib2", "HttpLib2Error")):
        try:
            mod = __import__(mod_name, fromlist=[cls_name])
            if isinstance(err, getattr(mod, cls_name)):
                return True
        except Exception:
            pass
    return False


def _classify_unavailable(err: Exception) -> Optional[str]:
    """Classify a decode error as a tolerable Google-unavailability condition
    ("server_unavailable" for a network error / HTTP 5xx, "quota_exceeded" for
    HTTP 429) or None when it must fail closed.
    """
    status = _http_status(err)
    if status == 429:
        return "quota_exceeded"
    if status is not None and 500 <= status <= 599:
        return "server_unavailable"
    if status is None and _is_network_error(err):
        return "server_unavailable"
    return None


def decrypt_and_verify_integrity_verdict(integrity_token: str) -> IntegrityVerdictResult:
    service_account_json = get_attestation_config().google_service_account_json
    if not service_account_json:
        log.error("GOOGLE_SERVICE_ACCOUNT_JSON not set")
        track_event(
            "AndroidAuth.PlayIntegrity.ConfigMissing", {"missing": "GOOGLE_SERVICE_ACCOUNT_JSON"}
        )
        return IntegrityVerdictResult.fail_hard("service_account_not_configured")
    try:
        credentials_info = json.loads(service_account_json)
    except Exception as err:
        track_exception(err, {"source": "decrypt_and_verify_integrity_verdict.parseCreds"})
        return IntegrityVerdictResult.fail_hard("service_account_parse_error")

    package_name = get_attestation_config().android_package_name
    if not package_name:
        track_event("AndroidAuth.PlayIntegrity.ConfigMissing", {"missing": "ANDROID_PACKAGE_NAME"})
        return IntegrityVerdictResult.fail_hard("package_name_not_configured")

    dep_start = time.time() * 1000
    try:
        from google.oauth2 import service_account
        from googleapiclient.discovery import build

        creds = service_account.Credentials.from_service_account_info(
            credentials_info,
            scopes=["https://www.googleapis.com/auth/playintegrity"],
        )
        service = build("playintegrity", "v1", credentials=creds, cache_discovery=False)

        log.info("Decrypting Play Integrity token...")
        response = (
            service.v1()
            .decodeIntegrityToken(packageName=package_name, body={"integrityToken": integrity_token})
            .execute()
        )
        dep_duration = time.time() * 1000 - dep_start

        verdict = response.get("tokenPayloadExternal")
        if not verdict:
            track_dependency(
                name="PlayIntegrity.decodeIntegrityToken",
                target="playintegrity.googleapis.com",
                data=f"packageName={package_name}",
                duration=dep_duration,
                success=False,
                result_code="empty-response",
                properties={"packageName": package_name},
            )
            return IntegrityVerdictResult.fail_hard("empty_response")

        track_dependency(
            name="PlayIntegrity.decodeIntegrityToken",
            target="playintegrity.googleapis.com",
            data=f"packageName={package_name}",
            duration=dep_duration,
            success=True,
            result_code=200,
            properties={
                "packageName": package_name,
                "appRecognition": (verdict.get("appIntegrity") or {}).get("appRecognitionVerdict"),
                "deviceRecognition": ",".join(
                    (verdict.get("deviceIntegrity") or {}).get("deviceRecognitionVerdict") or []
                ),
                "playProtect": (verdict.get("environmentDetails") or {}).get("playProtectVerdict"),
            },
        )
        return IntegrityVerdictResult.success(verdict)
    except Exception as err:
        dep_duration = time.time() * 1000 - dep_start
        log.error("Play Integrity API error: %s", err)
        track_dependency(
            name="PlayIntegrity.decodeIntegrityToken",
            target="playintegrity.googleapis.com",
            data=f"packageName={package_name}",
            duration=dep_duration,
            success=False,
            result_code="exception",
            properties={"packageName": package_name, "message": str(err)},
        )
        track_exception(err, {"source": "decrypt_and_verify_integrity_verdict", "packageName": package_name})
        kind = _classify_unavailable(err)
        return IntegrityVerdictResult.fail_tolerable(kind) if kind else IntegrityVerdictResult.fail_hard("api_error")

File: backend_samples/library/java/AzureAIVisionFaceDeviceAttestation/src/main/java/com/azure/ai/vision/face/deviceattestation/android/PlayIntegrityApi.java
package com.azure.ai.vision.face.deviceattestation.android;

import java.io.ByteArrayInputStream;
import java.nio.charset.StandardCharsets;
import java.util.List;

import com.azure.ai.vision.face.deviceattestation.Maps;
import com.azure.ai.vision.face.deviceattestation.config.AttestationConfig;
import com.azure.ai.vision.face.deviceattestation.logging.AttestationLogger;
import com.azure.ai.vision.face.deviceattestation.logging.DependencyTelemetry;
import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
import com.google.api.client.http.HttpResponseException;
import com.google.api.client.http.HttpTransport;
import com.google.api.client.json.gson.GsonFactory;
import com.google.api.services.playintegrity.v1.PlayIntegrity;
import com.google.api.services.playintegrity.v1.model.DecodeIntegrityTokenRequest;
import com.google.auth.http.HttpCredentialsAdapter;
import com.google.auth.oauth2.GoogleCredentials;

/**
 * The network/credentials boundary for Google's Play Integrity API. Reads the
 * service-account JSON + package name from config and decodes a client-supplied
 * integrity token into an {@link IntegrityVerdictResult}.
 */
public final class PlayIntegrityApi {

    private static final String SCOPE = "https://www.googleapis.com/auth/playintegrity";

    /**
     * Outcome of a Play Integrity decode attempt. {@code tolerable} is true only for
     * the Google-unavailable (network / HTTP 5xx) or quota-exceeded (HTTP 429)
     * conditions the {@code allowAndroidAttestationWhenGoogleUnavailable} policy may
     * accept; every other failure fails closed.
     */
    public static final class IntegrityVerdictResult {
        public final boolean ok;
        public final PlayIntegrityVerdict verdict;
        public final boolean tolerable;
        public final String reason;

        private IntegrityVerdictResult(boolean ok, PlayIntegrityVerdict verdict, boolean tolerable, String reason) {
            this.ok = ok;
            this.verdict = verdict;
            this.tolerable = tolerable;
            this.reason = reason;
        }

        static IntegrityVerdictResult success(PlayIntegrityVerdict verdict) {
            return new IntegrityVerdictResult(true, verdict, false, "ok");
        }

        static IntegrityVerdictResult failTolerable(String reason) {
            return new IntegrityVerdictResult(false, null, true, reason);
        }

        static IntegrityVerdictResult failHard(String reason) {
            return new IntegrityVerdictResult(false, null, false, reason);
        }
    }

    public static IntegrityVerdictResult decryptAndVerifyIntegrityVerdict(
            AttestationConfig config,
            AttestationLogger logger,
            String integrityToken) {
        String serviceAccountJson = config.googleServiceAccountJson();
        if (serviceAccountJson == null || serviceAccountJson.isEmpty()) {
            logger.trackEvent("AndroidAuth.PlayIntegrity.ConfigMissing", Maps.of("missing", "GOOGLE_SERVICE_ACCOUNT_JSON"), null);
            return IntegrityVerdictResult.failHard("service_account_not_configured");
        }
        String packageName = config.androidPackageName();
        if (packageName == null || packageName.isEmpty()) {
            logger.trackEvent("AndroidAuth.PlayIntegrity.ConfigMissing", Maps.of("missing", "ANDROID_PACKAGE_NAME"), null);
            return IntegrityVerdictResult.failHard("package_name_not_configured");
        }

        GoogleCredentials credentials;
        try {
            credentials = GoogleCredentials
                    .fromStream(new ByteArrayInputStream(serviceAccountJson.getBytes(StandardCharsets.UTF_8)))
                    .createScoped(List.of(SCOPE));
        } catch (Exception e) {
            logger.trackException(e, Maps.of("source", "decryptAndVerifyIntegrityVerdict.parseCreds"));
            return IntegrityVerdictResult.failHard("service_account_parse_error");
        }

        long start = System.nanoTime();
        try {
            HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
            PlayIntegrity service = new PlayIntegrity.Builder(transport, GsonFactory.getDefaultInstance(), new HttpCredentialsAdapter(credentials))
                    .setApplicationName("AzureAIVisionFaceDeviceAttestation")
                    .build();
            DecodeIntegrityTokenRequest request = new DecodeIntegrityTokenRequest().setIntegrityToken(integrityToken);
            var response = service.v1().decodeIntegrityToken(packageName, request).execute();
            long durationMs = (System.nanoTime() - start) / 1_000_000;

            if (response == null || response.getTokenPayloadExternal() == null) {
                logger.trackDependency(dependency(packageName, durationMs, false, "empty-response"));
                return IntegrityVerdictResult.failHard("empty_response");
            }

            PlayIntegrityVerdict verdict = map(response.getTokenPayloadExternal());
            logger.trackDependency(dependency(packageName, durationMs, true, "200"));
            return IntegrityVerdictResult.success(verdict);
        } catch (Exception e) {
            long durationMs = (System.nanoTime() - start) / 1_000_000;
            logger.trackDependency(dependency(packageName, durationMs, false, "exception"));
            logger.trackException(e, Maps.of("source", "decryptAndVerifyIntegrityVerdict", "packageName", packageName));
            String kind = classifyUnavailable(e);
            return kind != null ? IntegrityVerdictResult.failTolerable(kind) : IntegrityVerdictResult.failHard("api_error");
        }
    }

    /**
     * Classify a decode error as a tolerable Google-unavailability condition
     * ("server_unavailable" for a network error / HTTP 5xx, "quota_exceeded" for
     * HTTP 429) or null when it must fail closed.
     */
    private static String classifyUnavailable(Exception e) {
        if (e instanceof HttpResponseException hre) {
            int status = hre.getStatusCode();
            if (status == 429) {
                return "quota_exceeded";
            }
            if (status >= 500 && status <= 599) {
                return "server_unavailable";
            }
            return null;
        }
        if (e instanceof java.net.UnknownHostException
                || e instanceof java.net.ConnectException
                || e instanceof java.net.SocketTimeoutException
                || e instanceof java.net.NoRouteToHostException) {
            return "server_unavailable";
        }
        return null;
    }

    private static DependencyTelemetry dependency(String packageName, long durationMs, boolean success, String resultCode) {
        return DependencyTelemetry.builder("PlayIntegrity.decodeIntegrityToken")
                .target("playintegrity.googleapis.com")
                .data("packageName=" + packageName)
                .durationMillis(durationMs)
                .success(success)
                .resultCode(resultCode)
                .properties(Maps.of("packageName", packageName))
                .build();
    }

    private static PlayIntegrityVerdict map(com.google.api.services.playintegrity.v1.model.TokenPayloadExternal t) {
        PlayIntegrityVerdict verdict = new PlayIntegrityVerdict();

        var rd = t.getRequestDetails();
        if (rd != null) {
            PlayIntegrityVerdict.RequestDetails d = new PlayIntegrityVerdict.RequestDetails();
            d.requestPackageName = rd.getRequestPackageName();
            d.timestampMillis = rd.getTimestampMillis() != null ? String.valueOf(rd.getTimestampMillis()) : null;
            d.requestHash = rd.getRequestHash();
            verdict.requestDetails = d;
        }

        var ad = t.getAccountDetails();
        if (ad != null) {
            PlayIntegrityVerdict.AccountDetails d = new PlayIntegrityVerdict.AccountDetails();
            d.appLicensingVerdict = ad.getAppLicensingVerdict();
            verdict.accountDetails = d;
        }

        var ai = t.getAppIntegrity();
        if (ai != null) {
            PlayIntegrityVerdict.AppIntegrity d = new PlayIntegrityVerdict.AppIntegrity();
            d.appRecognitionVerdict = ai.getAppRecognitionVerdict();
            d.packageName = ai.getPackageName();
            d.certificateSha256Digest = ai.getCertificateSha256Digest();
            d.versionCode = ai.getVersionCode() != null ? String.valueOf(ai.getVersionCode()) : null;
            verdict.appIntegrity = d;
        }

        var di = t.getDeviceIntegrity();
        if (di != null) {
            PlayIntegrityVerdict.DeviceIntegrity d = new PlayIntegrityVerdict.DeviceIntegrity();
            d.deviceRecognitionVerdict = di.getDeviceRecognitionVerdict();
            var rda = di.getRecentDeviceActivity();
            if (rda != null) {
                PlayIntegrityVerdict.RecentDeviceActivity r = new PlayIntegrityVerdict.RecentDeviceActivity();
                r.deviceActivityLevel = rda.getDeviceActivityLevel();
                d.recentDeviceActivity = r;
            }
            verdict.deviceIntegrity = d;
        }

        var ed = t.getEnvironmentDetails();
        if (ed != null) {
            PlayIntegrityVerdict.EnvironmentDetails d = new PlayIntegrityVerdict.EnvironmentDetails();
            d.playProtectVerdict = ed.getPlayProtectVerdict();
            var arv = ed.getAppAccessRiskVerdict();
            if (arv != null) {
                PlayIntegrityVerdict.AppAccessRiskVerdict a = new PlayIntegrityVerdict.AppAccessRiskVerdict();
                a.appsDetected = arv.getAppsDetected();
                d.appAccessRiskVerdict = a;
            }
            verdict.environmentDetails = d;
        }

        return verdict;
    }

    private PlayIntegrityApi() {
    }
}
